### PR TITLE
DataPoint/pyeodh Integrations

### PR DESCRIPTION
**Pull request by dwest77a**
Starting point for ongoing integrations
 - Added `get_cloud_products` method for pyeodh Items.

### CODE REVIEW

**Comment by dwest77a** [comment]
Note: This Pull Request is not ready for merging and will require feedback from other developers.

**Comment by dwest77a** [comment]
DataPoint v0.4.1 has now been released for the purpose of integration with pyeodh

**Comment by dwest77a** [comment]
The QA checks now fail with `isort` but I can't see why. Is there some non-standard configuration being applied?

**Comment by figi44** [comment]
@dwest77a running `poetry run isort pyeodh` should fix this, then you can verify it locally by running `make check`.

**Comment by figi44** [comment]
Added a `make format` command to format the codebase, feel free to rebase onto main to leverage it.

**Comment by dwest77a** [comment]
Unsure of the current reasons for failure, it looks like an error in the tests which I thought was because I'd changed the pytest version, but I put that back in place with the same issue. Also tried to run these tests from a fresh copy of `pyeodh` main with the same set of errors:

```
FAILED tests/ades_test.py::test_get_processes - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/ades_test.py::test_get_non_existent_process - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/ades_test.py::test_process_execution - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/ades_test.py::test_get_jobs - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/ades_test.py::test_get_job - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/ades_test.py::test_delete_job - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/wmts_test.py::test_get_wmts - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/wmts_test.py::test_get_wms - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/workspace_test.py::test_upload_file[test content-test.txt] - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
FAILED tests/workspace_test.py::test_upload_file[tests/data/test.txt-test.txt] - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_catalog_service - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_ping - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_conformance - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_catalog - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_catalogs_from_catalog - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collections_from_catalog - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collection_from_catalog - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collection_items - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collection_items_with_limit - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collection_items_total_count - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_collection_item - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_get_item_from_href - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
ERROR tests/resource_catalog_test.py::test_conformance_error_raised - AttributeError: 'VCRHTTPResponse' object has no attribute 'version_string'
```

**Comment by figi44** [comment]
This was a bug in vcrpy 6.0.2 (i guess poetry updated it for you from 6.0.1). I updated it to 7.0.0 in the main branch, which also required regenerating vcrs. Please rebase onto main, you'll need to run `poetry lock` and `poetry install` afterwards. Then you can verify the tests by running `make test`

**Comment by dwest77a** [comment]
@figi44 This should now be ok, I'll add a test to cover the non-covered section so all tests pass, and update the README with a few extra details.

**Comment by dwest77a** [comment]
@figi44 I think this should be fine now, I had been adding a test to the `test_conformance_error_raised` which only tested the conformance error, like with the item delete functionality. Instead I've now added a new test case (function?) called `test_get_cloud_products` which does test the new features, and seems to pass all checks here.

Do I also need the conformance error part or is that not needed? (I've kept the conformance error lines in my function and the test_resource_catalog script, just commented out, as it was causing the other test case to fail). Thanks

**Comment by dwest77a** [comment]
I believe the minimum coverage limit was reduced, so this branch is good to go with the merge when possible @figi44, this is directly relevant for the User Workshop on the 26th Feb, and will feed into the `eodh_training` package as a new pyeodh version. Once this is merged and a new release is published, I can update the training package requirements with the correct pyeodh version. Thanks

File: pyeodh/resource_catalog.py [hunk 1]
```diff
@@ -115,6 +115,67 @@ def update(
...
+        Added feature that uses the CEDA
+        DataPoint library to create a list of
+        CloudProduct objects here."""
+
+        # if not self.get_root().check_conforms_to(
+        #    Conformance.TRANSACTION_EXTENSION.value
+        # ):
+        #    raise ConformanceError(
+        #        f"{Conformance.TRANSACTION_EXTENSION.value}",
+        #    )
```
Comment (figi44) [inline pyeodh/resource_catalog.py:131]: I don't think we need a conformance check, unless there is a conformance class we need the API to support in order for this feature to work.

File: pyeodh/resource_catalog.py [hunk 2]
```diff
@@ -115,6 +115,67 @@ def update(
...
+                DataPointCloudProduct,
+                DataPointCluster,
+            )
+            from ceda_datapoint.core.item import identify_cloud_type
+        except ImportError:
+            logger.error(
+                "Unable to import DataPoint connectors - install "
+                " with `pip install ceda-datapoint`"
+            )
+            return None
```
Comment (figi44) [inline pyeodh/resource_catalog.py:144]: Why not import this at the top of the module when `ceda-datapoint` is set as a dependency in `pyproject.toml`?

File: pyeodh/resource_catalog.py [hunk 3]
```diff
@@ -115,6 +115,67 @@ def update(
         if resp_data:
             self._set_props(self._pystac_object.from_dict(resp_data))
 
+    def get_cloud_products(
+        self,
+    ) -> object | None:
```
Comment (figi44) [inline pyeodh/resource_catalog.py:120]: Can we be more specific here please? Looks like we return either `DataPointCloudProduct`, `DataPointCluster`, an empty list? or `None`. I think this makes the method quite unpredictable and hard to handle by its consumers.

File: pyeodh/resource_catalog.py [hunk 4]
```diff
@@ -115,6 +115,67 @@ def update(
...
+        """
+        Added feature that uses the CEDA
+        DataPoint library to create a list of
+        CloudProduct objects here."""
```
Comment (figi44) [inline pyeodh/resource_catalog.py:124]: Please follow the google docstring style, see other methods or https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings for reference.

**Review by figi44 — CHANGES_REQUESTED** [review]

File: pyeodh/resource_catalog.py (hunk 1 above)
Comment (dwest77a) [reply pyeodh/resource_catalog.py]: Removed this section

File: pyeodh/resource_catalog.py (hunk 2 above)
Comment (dwest77a) [reply pyeodh/resource_catalog.py]: Moved these to the top without import check, also removed the import test (required to increase coverage above minimum limit)

File: pyeodh/resource_catalog.py (hunk 3 above)
Comment (dwest77a) [reply pyeodh/resource_catalog.py]: Added the possible options for return values, no longer returns None with import failure as this section has been removed.

File: pyeodh/resource_catalog.py (hunk 4 above)
Comment (dwest77a) [reply pyeodh/resource_catalog.py]: Updated the docstring

**Comment by figi44** [comment]
Thanks @dwest77a!
